### PR TITLE
SAK-30325 send user direct link instead of proxying data from s3/swift

### PR DESCRIPTION
**Pull request by ottenhoff**
Don't proxy large files into JVM memory to deliver them to user.... two main things here:

a) Write to a temp file if the blob asset is too large.

b) Send the user a direct link to the asset (usually a pre-signed, temporary link) instead of forcing the asset to be proxied by the Sakai JVM

### CODE REVIEW

File: cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/SwiftFileSystemHandler.java [hunk 1]
```diff
@@ -266,6 +274,21 @@ public void destroy() throws IOException{
...
+	public URI getAssetDirectLink(String id, String root, String filePath) throws IOException {
+        ContainerAndName can = getContainerAndName(id, root, filePath);
+        ObjectApi objectApi = swiftApi.getObjectApi(region, can.container);
+        SwiftObject so = objectApi.get(can.name, GetOptions.NONE);
+        if(so == null){
+            throw new IOException("No object found for " + id);
+        }
+        
+        return so.getUri();
+	}
```
Comment (botimer) [inline cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/SwiftFileSystemHandler.java]: Tabs vs. spaces

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java [hunk 2]
```diff
@@ -7088,9 +7111,9 @@ protected void handleAccessResource(HttpServletRequest req, HttpServletResponse
 		        } 
 		        else 
 		        {
-		        	// Output partial content. Adapted from Apache Tomcat 5.5.27 DefaultServlet.java
-		        	
-		        	res.setStatus(HttpServletResponse.SC_PARTIAL_CONTENT);
+		            // Output partial content. Adapted from Apache Tomcat 5.5.27 DefaultServlet.java
+		            res.addHeader("Accept-Ranges", "bytes");
+		            res.setStatus(HttpServletResponse.SC_PARTIAL_CONTENT);
```
Comment (botimer) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: Mixed indentation

**Comment by botimer** [comment]
This look good to me and tests out well so far. I've made comments about indentation in two places but, otherwise, nothing more to say.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/DbContentService.java [hunk 3]
```diff
@@ -2307,6 +2308,25 @@ public InputStream streamResourceBody(ContentResource resource) throws ServerOve
...
         }
+        
+        /**
+         * Return a URI containing a direct link to the asset.
+         * 
+         * @param rootFolder
+         * @param resource
+         * @return URI representing a direct link to the asset or null
+         */
+        public URI getDirectLink(ContentResource resource)
```
Comment (ern) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/DbContentService.java]: Should this method be in BaseContentService?

File: cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java [hunk 4]
```diff
@@ -182,15 +234,73 @@ public void destroy() throws IOException{
...
+        if (temporaryBlobDirectory != null) {
+            check = new File(temporaryBlobDirectory, filename);
+        }
+        else {
+            check = new File(System.getProperty("java.io.tmpdir"), filename);
+        }
+
+        if (check.exists()) {
+            try {
+                stream = new FileInputStream(check);
```
Comment (ern) [inline cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java]: Would use the new try-with-resources here:
try (stream = new FileInputStream(check)) {

File: cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java [hunk 5]
```diff
@@ -182,15 +234,73 @@ public void destroy() throws IOException{
...
+            } catch (FileNotFoundException e) {
+                stream = null;
+            }
+        } else {
+            try {
+                FileOutputStream fos = new FileOutputStream(check);
```
Comment (ern) [inline cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java]: same as above
try (FileOutputStream fos = new FileOutputStream(check)) {

**Comment by ern** [comment]
I would also recommend adding a Vector<String, Date> to keep track of the files in the temp dir and clean up them up every interval and remove files past after a certain time. Also remember when the thread shuts down it should remove everything!

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/DbContentService.java (hunk 3 above)
Comment (botimer) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/DbContentService.java]: There is a case to be made, but as it stands, the FileSystemHandler is only at the DbContentService. I don't think we want to move it to the interface, so I see this as internal to this class.

**Comment by botimer** [comment]
I'm not so sure about tracking temp files in the thread/JVM. I think they should actually persist for a while to avoid redownloading from the backend. It's worth mentioning that the temp files will likely only be used where the contents are needed for something other than a user download (say, copying a file). And, in those cases, only for files larger than the threshold (since the streams appear to get flaky in initial testing).

In some ways, the temp files are superstition to begin with, and should be relatively rare. They will act as an external cache, so I'm inclined to say that external cleanup (cron/quartz) is the best recommendation.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java [hunk 6]
```diff
@@ -13431,6 +13454,13 @@ public void setOldDisplayName(String oldDisplayName) {
 		public void open();
 
 		/**
+		 * Get a direct link to the asset so it doesn't have to be streamed.
+		 * @param resource
+		 * @return URI or null if no direct link is available
+		 */
+		public URI getDirectLink(ContentResource resource);
```
Comment (buckett) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: Doesn't allow it to work on earlier JVMs but I wonder for public APIs like this if we should start returning `Optional<URI>` to make it clearer the caller should check to see if there is a value present.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java (hunk 2 above)
Comment (ottenhoff) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: Yeah but this whole section of code has odd spacing/tabs.  The odd combination shown here is consistent with the surrounding code.

File: cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java (hunk 4 above)
Comment (ottenhoff) [inline cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java]: Except we're trying to send an open stream here ..... we don't want this stream auto-closed

File: cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java (hunk 5 above)
Comment (ottenhoff) [inline cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java]: FileCopyUtils responsible for closing both streams

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java (hunk 6 above)
Comment (ottenhoff) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: I like the Optional<URI> suggestion a lot .... I'll do it as a sub-JIRA once I'm satisfied this code works well on Java 7 .....

**Comment by ottenhoff** [comment]
Okay I fixed one tab vs space issue...

File: cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java (hunk 4 above)
Comment (ern) [inline cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java]: right good point!

File: cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java (hunk 5 above)
Comment (ern) [inline cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java]: OK, i probably should have opened the source and looked!

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java (hunk 6 above)
Comment (ern) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: Yeah this makes a lot of sense, though I think we should officially mention when Java 8 syntax can be used.

**Comment by ern** [comment]
@botimer On the temp file issue I am fine with a destroy method that removes the temp files. I think keeping them around unattended is not ideal as they will be large. Maybe we don't need a thread to manage it but it does make sense to remove the files when the app is shutdown.

**Comment by ottenhoff** [comment]
@ern I agree with @botimer here.... we want the files to live on beyond a Sakai restart. A restart of Tomcat does nothing to nullify the validity of the file

**Comment by ottenhoff** [comment]
@botimer I had to re-add the blob => into memory => bytearray stream as the direct blob stream doesn't work (e.g., Lessons resources)

I did it as a separate commit .... happy to squash before merging .....

**Comment by ern** [comment]
@ottenhoff @botimer wdyt about not using java.io.tempdir then and using path somewhere in $SAKAI_HOME? My only concern is leaving these files somewhere where nobody expects them?

java.io.tmpdir usually defaults to /tmp which is a typically a tmpfs which usually is a memory filesystem though I do like how this is auto cleaned up by the OS on reboot its probably not wise to store large files like this there, comments?

**Comment by botimer** [comment]
Hmm.. Still pretty strange that we can't count on the streams. I haven't been able to determine what's causing them to be unstable. Kind of a bummer for larger files, but at least this will only happen on internal things if the signed URLs are in use.

There might be some thrashing for some copy operation that reads a bunch of files, but the byte arrays would be short lived with almost no reference count, and easily cleaned up. It shouldn't be too big a problem, provided there isn't a ton of concurrent stuff against big files soaking up too much heap.

It's definitely worth moving this forward, though, because it's a lot better behaved than what's in master at the moment.

**Comment by botimer** [comment]
@ern A good idea, but I think we've got the probabilistic best option in place.

We're allowing it to be configured, and fall back to the JVM temp dir. I think this is pretty advanced (and rare to date) deployment stuff, so that's probably the best approach. Tomcat, by default, sets its own temp directory. That can be configured at the Tomcat level, or you can override at the Sakai level.

I'm not sure there is much to be gained by inventing something else in sakai.home, to be honest. If you want the files put somewhere special, you have two ways to do that. I think it's fair to expect a deployer to consider this question and we probably can't make a much better default than we have.

As we gain experience, a better approach may emerge.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java [hunk 7]
```diff
@@ -7015,11 +7015,42 @@ protected void handleAccessResource(HttpServletRequest req, HttpServletResponse
 					return; 
 				}
 
+				// If there is a direct link to the asset, no sense streaming it.
+				// Send the asset directly to the load-balancer or to the client
+				URI directLinkUri = m_storage.getDirectLink(resource);
```
Comment (ottenhoff) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: So if (resource instanceof WrappedContentResource) the rest of the code below will break .....

@botimer or @buckett any hints why?

Comment (botimer) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: The only thing that comes to mind is the content filter stuff having set a conflicting header. It's safe to overwrite header values, so this is a long shot.

It looks like the wrapper (resource and stream) handles things gracefully. If you have some repro steps, I can fire up a debugger.

Comment (ottenhoff) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: Basic repo steps:

Create an HTML file in resources and paste 10 paragraphs or so of lipsum text ...... Then try to view file ..... experiment with enabling/disabling wrapping the content via Actions -> Edit Details

Comment (buckett) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: @ottenhoff What do you mean by break?

The Content-Length header will be wrong if you're not reading from the InputStream and aren't redirecting them directly to the content. As with a WrappedContentResource extends the content by adding on a HTML header/footer which means the content length needs increasing. I think a browser will hang around waiting for the extra data which is never going to come.

Comment (botimer) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java]: We just fixed this up. There was a cast to BaseResourceEdit failing and getting translates into a silent 404. I think Sam will update the pull request today.

**Comment by ottenhoff** [comment]
rebased and pushed. thanks @botimer 

happy to squash down to one commit if desired

**Comment by botimer** [comment]
I re-ran the Travis build to get past the unrelated search issue and all is now well. I think this is ready for merge.

**Comment by ottenhoff** [comment]
I haven't been following the merging conversations.... is there a desire to squash this all the way down into one commit? or are the three commits okay?

**Comment by botimer** [comment]
I'm indifferent here. Each commit is tagged with the issue, so whatever works for others is fine for me.

**Comment by buckett** [comment]
Is this ok to go in now?

**Comment by ottenhoff** [comment]
Yes. Squash or no-squash?

**Pull request merged**